lib/elixir/lib/calendar/datetime.ex

```diff
@@ -93,6 +93,131 @@ defmodule DateTime do
     System.os_time() |> from_unix!(:native, calendar)
   end
 
+  @doc """
+  Builds a datetime from date and time structs
+
+  It expects a time zone to put the `DateTime` in.
+  If the time zone is not passed it will default to "Etc/UTC",
+  which always succeeds. Otherwise, the DateTime is checked against the time zone database
+  given as `time_zone_database`. See the "Time zone database"
+  section in the module documentation.
+
+  ## Examples
+
+      iex> DateTime.new(~D[2016-05-24], ~T[13:26:08.003], "Etc/UTC")
+      {:ok, ~U[2016-05-24 13:26:08.003Z]}
+
+  When the datetime is ambiguous - for instance during changing from summer
+  to winter time - the two possible valid datetimes are returned. First the one
+  that happens first, then the one that happens after.
+
+      iex> {:ambiguous, first_dt, second_dt} = DateTime.new(~D[2018-10-28], ~T[02:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> first_dt
+      #DateTime<2018-10-28 02:30:00+02:00 CEST Europe/Copenhagen>
+      iex> second_dt
+      #DateTime<2018-10-28 02:30:00+01:00 CET Europe/Copenhagen>
+
+  When there is a gap in wall time - for instance in spring when the clocks are
+  turned forward - the latest valid datetime just before the gap and the first
+  valid datetime just after the gap.
+
+      iex> {:gap, just_before, just_after} = DateTime.new(~D[2019-03-31], ~T[02:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> just_before
+      #DateTime<2019-03-31 01:59:59.999999+01:00 CET Europe/Copenhagen>
+      iex> just_after
+      #DateTime<2019-03-31 03:00:00+02:00 CEST Europe/Copenhagen>
+
+  Most of the time there is one, and just one, valid datetime for a certain
+  date and time in a certain time zone.
+
+      iex> {:ok, datetime} = DateTime.new(~D[2018-07-28], ~T[12:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> datetime
+      #DateTime<2018-07-28 12:30:00+02:00 CEST Europe/Copenhagen>
+  """
+  @doc since: "1.11.0"
+  @spec new(Date.t(), Time.t(), Calendar.time_zone(), Calendar.time_zone_database()) ::
+          {:ok, t}
+          | {:ambiguous, t, t}
+          | {:gap, t, t}
+          | {:error,
+             :incompatible_calendars | :time_zone_not_found | :utc_only_time_zone_database}
+  def new(
+        date,
+        time,
+        time_zone \\ "Etc/UTC",
+        time_zone_database \\ Calendar.get_time_zone_database()
+      )
+
+  def new(date, time, time_zone, time_zone_database) do
+    with {:ok, naive_datetime} <- NaiveDateTime.new(date, time) do
+      from_naive(naive_datetime, time_zone, time_zone_database)
```

Comment:
to simplify this I just decided to create a `NaiveDateTime` and use the already existing `from_naive` to accomplish this. If we would prefer not to do this let me know, I was trying to limit the repetition this could cause

Reply:
In this case, we want to avoid the double conversion. I think in this case what you can do is to:
 
1. rewrite `from_naive` as `new/4`
2. `from_naive` is the same as calling `new(naive_datetime, naive_datetime, time_zone, time_zone_database)`. 

This way we keep the code simple (no duplication) and fast (no converting to naive datetime and then datetime). WDYT?

Reply:
@josevalim Thanks for reviewing this so quick! I have a few questions/comments

 I ran into an issue where `Calendar.TimeZoneDatabase.time_zone_periods_from_wall_datetime` expects `(Calendar.naive_datetime(), Calendar.time_zone())`https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/calendar/time_zone_database.ex#L66 
So don't I need to create a `NaiveDatetime` in order to determine if it's ambiguous or has a gap?

I understand your second point that makes sense that `from_naive(naive_datetime)` is the same as `new(naive_datetime, naive_datetime)`

So is point 1) mean you want to rewrite `from_naive` to use `new/4` internally? Or that we should rewrite `new/4` using the same functions as `from_naive`? If you could clarify this a little that would be great!

Reply:
I see your point. Given a naive_datetime is required anyway, the current code (as is) is the best way to go.

Perhaps my only suggestion is to add a clause like this:

```elixir
  def new(%Date{calendar: calendar} = date, %Time{calendar: calendar} = time, "Etc/UTC", _db) do
    %{year: year, month: month, day: day} = date
    %{hour: hour, minute: minute, second: second, microsecond: microsecond} = time

    datetime = %DateTime{
      calendar: calendar,
      year: year,
      month: month,
      day: day,
      hour: hour,
      minute: minute,
      second: second,
      microsecond: microsecond,
      std_offset: 0,
      utc_offset: 0,
      zone_abbr: "UTC",
      time_zone: "Etc/UTC"
    }

    {:ok, datetime}
  end
```

So we at least optimize the most common scenario (UTC). WDYT?

Reply:
Okay added this clause!